### ISSUE TITLE
WASI threads support

### ISSUE THREAD

**Issue by seclerp** [extism/extism#357]
I'm currently working on a .NET PDK prototype. NativeAOT-LLVM toolchain that I'm using generate WASI-compliant, completely AOTed `.wasm` binaries very well, and it also uses cutting-edge features like `wasi-threads`.

Wasmer is working great with such kind of binaries, but Wasmtime struggles a lot and requires additional configuration.

1. The first issue that I encouraged is the missing wasm-threads support (this and below errors are from the .NET host):
```
failed to parse WebAssembly module

Caused by:
    Invalid input WebAssembly code at offset 1516: threads must be enabled for shared memories
```

I created [an Extism fork](https://github.com/extism/extism/compare/main...seclerp:extism:wasm-threads) that adds necessary flag to the Extism binaries for .NET SDK. (BTW it is published [here](https://www.nuget.org/packages/Extism.runtime.wasm-threads.win-x64/))

2. The first issue has gone, but there is the second one:
```
unknown import: `env::memory` has not been defined
```

After investigation I found that thread spawning and thread `memory` import should be done on the Extism side. There is a crate [`wasmtime-wasi-threads`](https://docs.rs/wasmtime-wasi-threads/latest/wasmtime_wasi_threads/) that adds missing puzzle pieces but it requires the codebase to be modified in a thread-safe manner.

I'm not an expert in Rust so I've tried to adopt the codebase around internal.rs and plugin.rs a bit but don't have a lot of time to continue, but I think you folks could do it quite quickly. The main pain points for me are raw pointers that are hard to wrap in `Arc<Mutex<>>` or something like that in a performant manner.

Hope that you folks have some plans in this direction.

**Comment by syrusakbary** [comment]
I propose a solution: why not use Wasmer? It would help for the threads issue, but also for Extism target the browser without reimplementing it :)

**Comment by nilslice** [comment]
@syrusakbary - would you be open to helping out with a scan of the Extism runtime codebase? There are some places where we do tie fairly closely to the Wasmtime APIs & some audit of what Wasmer could simplify or may not have consistent support would be really helpful. 

We agree it would be ideal to re-use the whole runtime for the browser, vs. the re-implementation approach. And we also would be happy to have support for Wasmer as an alternative implementation, just as we're considering Wazero for native Go hosts.

@seclerp - this reminds me to check in about https://github.com/extism/extism/pull/362 -- is that still progressing? Do you need any help? Whenever you're able to, please feel free to reach out!

**Comment by syrusakbary** [comment]
From looking at the runtime source code, it should be fairly simple:

https://github.com/extism/extism/blob/main/runtime/

The main implementation differences would be:
1. Moving from a "linker" approach (wasmtime) into an "imports" approach (wasmer)
2. Move from shared info in `Context` (now stored by the linker), to a Wasmer Environment
3. `wasi_nn` is not yet supported by Wasmer

The API is fairly similar, so in general, the changes would not be very hard.

In general, I think in max one or two days you can get everything ported. Including the same implementation working on the browser.

> as we're considering Wazero for native Go hosts

Wazero is awesome, so I think this also would make the most sense in go environments

**Comment by seclerp** [comment]
Hi folks, currently I have no free capacity but am still interested in implementing this in the near future - as a side project I'm working on .NET bindings for modern Wasmer, so I will be in the context of Wasmer implementation as well.

**Comment by elwerene** [comment]
it looks like wasmtime [added support to threads in the release 20.0.0](https://github.com/bytecodealliance/wasmtime/releases/tag/v20.0.0), so maybe this is something which might come soon to extism.

**Comment by bhelx** [comment]
We shipped wasmtime 20 support in https://github.com/extism/extism/releases/tag/v1.4.0

@seclerp could you test your code with the latest? No pressure if you are busy or not working on it anymore. Might be good to have @mhmd-azeez take a look at it.

**Comment by seclerp** [comment]
I have no time to work on this right now, sorry folks. Please feel free to pick up my changes for further use

**Comment by bhelx** [comment]
@seclerp no problem thank you for letting us know! Will have @mhmd-azeez take a look.

**Comment by mhmd-azeez** [comment]
@seclerp your .NET PDK prototype public? If yes, can you please share the link for it?

**Comment by seclerp** [comment]
Nope, I abandoned it once I saw your work on https://github.com/extism/dotnet-pdk. Anyway, it was on the stage of experiments at that moment.
